x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateMetricFieldMapper.java

```diff
@@ -0,0 +1,339 @@
+
+    private Explicit<Set<String>> metrics;
+
+    private AggregateMetricFieldMapper(String simpleName, MappedFieldType fieldType, MappedFieldType defaultFieldType,
+        Settings indexSettings, MultiFields multiFields, Explicit<Boolean> ignoreMalformed, Explicit<Set<String>> metrics,
+        CopyTo copyTo) {
+        super(simpleName, fieldType, defaultFieldType, indexSettings, multiFields, copyTo);
+        this.ignoreMalformed = ignoreMalformed;
+        this.metrics = metrics;
+    }
+
+    @Override
+    public AggregateMetricFieldType fieldType() {
+        return (AggregateMetricFieldType) super.fieldType();
+    }
+
+    @Override
+    protected String contentType() {
+        return fieldType.typeName();
+    }
+
+    @Override
+    protected AggregateMetricFieldMapper clone() {
+        return (AggregateMetricFieldMapper) super.clone();
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context, List<IndexableField> fields) throws IOException {
+        Object metricsAsObject;
+        if (context.externalValueSet()) {
+            metricsAsObject = context.externalValue();
+        } else {
+            metricsAsObject = context.parser().map();
+        }
+
+        if (metricsAsObject == null) {
+            metricsAsObject = fieldType().nullValue();
+        }
+
+        if (metricsAsObject == null) {
+            return;
+        }
+
+        Map<String, Object> parsedMetricsMap = new LinkedHashMap<>();
+        if (metricsAsObject instanceof Map) {
+            @SuppressWarnings("unchecked")
+            Map<String, Object> m = (Map<String, Object>) metricsAsObject;
+            for (Map.Entry<String, Object> e : m.entrySet()) {
+                String metricName = e.getKey();
+
+                if (metrics.value().contains(metricName) == false) {
+                    if (ignoreMalformed.value() == false) {
+                        throw new IllegalArgumentException("Aggregate metric [" + metricName +
+                            "] does not exist in the mapping of field [" + fieldType.name() + "]");
+                    } else {
+                        context.addIgnoredField(fieldType.name());
+                        continue;
+                    }
+                }
+
+                // Make sure that the value is a number. Probably this will change when
+                // new aggregate metric types are added (histogram, cardinality etc)
+                if (e.getValue() instanceof Number == false) {
+                    if (ignoreMalformed.value() == false) {
+                        throw new IllegalArgumentException("Aggregate metric [" + metricName +
+                            "] of field [" + fieldType.name() + "] must be a number");
+                    } else {
+                        context.addIgnoredField(fieldType.name());
+                        continue;
+                    }
+                }
+
+                // Make sure that value_count is not a negative value
```

Comment:
I think this comment is outdated... looks like the negative check is a bit further down

Reply:
fixed